internal/controller/notificationservice_controller.go

```diff
@@ -70,6 +71,7 @@ func (r *NotificationServiceReconciler) Reconcile(ctx context.Context, req ctrl.
 		err = AddFinalizerToPipelineRun(ctx, pipelineRun, r, NotificationPipelineRunFinalizer)
```

Comment:
why do we need all the calls to `IsAnnotationExistInPipelineRun` ? the predicate you've added should replace them.